OpenEdXMobile/src/main/java/org/edx/mobile/view/adapters/CourseSubSectionAdapter.kt

```diff
@@ -0,0 +1,228 @@
+                binding.bulkDownloadLayout.noOfVideos.text = "$totalDownloadableVideos"
+                val downloadedCount: Int = dbStore.getDownloadedVideosCountForSection(
+                    component.courseId,
+                    chapterId, sequentialId, null
+                )
+                if (downloadedCount == totalDownloadableVideos) {
+                    binding.bulkDownloadLayout.noOfVideos.setVisibility(true)
+                    setRowStateOnDownload(this, DownloadedState.DOWNLOADED, null)
+                } else if (dbStore.getDownloadingVideosCountForSection(
+                        component.courseId, chapterId,
+                        sequentialId, null
+                    ) + downloadedCount == totalDownloadableVideos
+                ) {
+                    binding.bulkDownloadLayout.noOfVideos.setVisibility(true)
+                    setRowStateOnDownload(
+                        this, DownloadedState.DOWNLOADING
+                    ) { itemClickListener.viewDownloadsStatus() }
+                } else {
+                    binding.bulkDownloadLayout.noOfVideos.setVisibility(true)
+                    setRowStateOnDownload(
+                        this, DownloadedState.ONLINE
+                    ) {
+                        val downloadableVideos: List<VideoBlockModel> =
+                            component.getVideos(true).filterIsInstance<VideoBlockModel>()
+                        for (videoBlockModel in downloadableVideos) {
+                            /**
+                             * Assign preferred downloadable url to [VideoBlockModel.downloadUrl],
+                             * to use this url to download. After downloading only downloaded
+                             * video path will be used for streaming.
+                             */
+                            /**
+                             * Assign preferred downloadable url to [VideoBlockModel.downloadUrl],
+                             * to use this url to download. After downloading only downloaded
+                             * video path will be used for streaming.
+                             */
+                            videoBlockModel.downloadUrl =
+                                VideoUtil.getPreferredVideoUrlForDownloading(videoBlockModel.data)
+                        }
+                        itemClickListener.download(downloadableVideos = downloadableVideos)
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Makes various changes to the row based on a video element's download status
+     *
+     * @param holder      ViewHolder of the row view
+     * @param state    current state of video download
+     * @param listener the listener to attach to the video download button
+     */
+    private fun setRowStateOnDownload(
+        holder: SubSectionViewHolder,
+        state: DownloadedState,
+        listener: View.OnClickListener?
+    ) {
+        when (state) {
```

Comment:
Use a root level `apply` on `holder.binding.bulkDownloadLayout` and place the `when` inside it

Reply:
replaced.